wpep0003.md

```diff
@@ -0,0 +1,142 @@
+| Key | Value |
+| ---: | :--- |
+| **Title** | Improve acl system scalibility |
+| **Number** | 0003 |
+| **Date** | 2019-12-10 |
+| **Authors** | Mehdi Abaakouk <mabaakouk@wazo.io> |
+| **Status** | proposed |
+
+## Abstract
+
+Redesign the way services checks token, alcs and tenant ownership
+
+## Motivation
+
+Currently Wazo-auth is heavily solicitated just for checking token validity, acls and tenant ownership relative to the supplied token.
+
+### First issue, A service do three API calls to wazo-auth before it starts to handle a request.
```

Comment:
The three requests is more a corner case than the most used
When the tenant from the header (or not header) match the tenant from token, no `GET /tenants` is triggered
The *normal* behavior only two requests because we should not cross tenant (but yeah I see the point with issue two)

Reply:
It's not a corner case a lot of endpoints uses `GET /tenants` to ensure tenant isolation.

If you pick confd:
* for GET ONE and PUT here: https://github.com/wazo-platform/wazo-confd/blob/master/wazo_confd/helpers/restful.py#L174
* for GET ALL when recurse=True here: https://github.com/wazo-platform/wazo-confd/blob/master/wazo_confd/helpers/restful.py#L78
And all confd resources use this helper.

Same for chatd: 
* https://github.com/wazo-platform/wazo-chatd/blob/master/wazo_chatd/plugins/presences/http.py#L43
* https://github.com/wazo-platform/wazo-chatd/blob/master/wazo_chatd/plugins/presences/http.py#L50
* https://github.com/wazo-platform/wazo-chatd/blob/master/wazo_chatd/plugins/presences/http.py#L24

Same for dird: 
https://github.com/wazo-platform/wazo-dird/blob/master/wazo_dird/plugins/phonebook/http.py#L43
https://github.com/wazo-platform/wazo-dird/blob/master/wazo_dird/plugins/displays/http.py#L51
https://github.com/wazo-platform/wazo-dird/blob/master/wazo_dird/plugins/displays/http.py#L58
https://github.com/wazo-platform/wazo-dird/blob/master/wazo_dird/plugins/displays/http.py#L65
https://github.com/wazo-platform/wazo-dird/blob/master/wazo_dird/plugins/displays/http.py#L28
...

Same for webhookd:
* https://github.com/wazo-platform/wazo-webhookd/blob/master/wazo_webhookd/plugins/subscription/http.py#L60
* https://github.com/wazo-platform/wazo-webhookd/blob/master/wazo_webhookd/plugins/subscription/http.py#L68
* https://github.com/wazo-platform/wazo-webhookd/blob/master/wazo_webhookd/plugins/subscription/http.py#L140

Reply:
> It's not a corner case a lot of endpoints uses `GET /tenants` to ensure tenant isolation.
> 
> If you pick confd:
> 
>     * for GET ONE and PUT here: https://github.com/wazo-platform/wazo-confd/blob/master/wazo_confd/helpers/restful.py#L174
> 
You right, but for non service user, the `GET /tenants` will trigger a 401 (which should be really fast).
Another solution to avoid breaking current logic is to refactor the logic behind this
 * to try to fetch resource with the tenant token first and if not found, try with `GET /tenants`
 * Or improve the 401 return of `GET /tenants`
 * Also focus to apply the proposition chosen on user resources first (administration second))

>     * for GET ALL when recurse=True here: https://github.com/wazo-platform/wazo-confd/blob/master/wazo_confd/helpers/restful.py#L78
>       And all confd resources use this helper.
> 
The GET ALL with recurse is generally used between service. But never used by the user/admin (all depends how your build your interface)

[this text only apply to proposition of reducing the request number (not removing them)]
It's quit hard without metric to view the real impact. Also it's subjective to associate the number of requests proportional to the load. I mean a `HEAD /token?scope` should not be the same load than a `GET /tenants`. Which request should be optimized? How many requests should be done? Why doing 3 requests on wazo-auth take more load than 100 requests on ARI? Reducing the number of requests of wazo-auth is a way to reduce load, but definitively not the only one, and if reducing the request by 3 to 1 (even if you multiply by 1000 on the same time) solve our problem than I fear the fragility of the wazo-auth